src/dimension_vector.c

```diff
@@ -0,0 +1,123 @@
+#include "dimension_vector.h"
+
+static int
+cmp_slices(const void *left, const void *right)
+{
+	const DimensionSlice *left_slice = *((DimensionSlice **) left);
+	const DimensionSlice *right_slice = *((DimensionSlice **) right);
+
+	if (left_slice->fd.range_start == right_slice->fd.range_start)
+	{
+		if (left_slice->fd.range_end == right_slice->fd.range_end)
+			return 0;
+
+		if (left_slice->fd.range_end > right_slice->fd.range_end)
+			return 1;
+
+		return -1;
+	}
+
+	if (left_slice->fd.range_start > right_slice->fd.range_start)
+		return 1;
+
+	return -1;
+}
+
+static int
+cmp_coordinate_and_slice(const void *left, const void *right)
+{
+	int64		coord = *((int64 *) left);
+	const DimensionSlice *slice = *((DimensionSlice **) right);
+
+	if (coord < slice->fd.range_start)
+		return -1;
+
+	if (coord >= slice->fd.range_end)
+		return 1;
+
+	return 0;
+}
+
+static DimensionVec *
+dimension_vec_expand(DimensionVec *vec, int32 new_capacity)
+{
+	if (vec != NULL && vec->capacity >= new_capacity)
+		return vec;
+
+	if (NULL == vec)
+		vec = palloc(DIMENSION_VEC_SIZE(new_capacity));
+	else
+		vec = repalloc(vec, DIMENSION_VEC_SIZE(new_capacity));
+
+	vec->capacity = new_capacity;
+
+	return vec;
+}
+
+DimensionVec *
+dimension_vec_create(int32 initial_num_slices)
+{
+	DimensionVec *vec = dimension_vec_expand(NULL, initial_num_slices);
+
+	vec->capacity = initial_num_slices;
+	vec->num_slices = 0;
+	return vec;
+}
+
+DimensionVec *
+dimension_vec_add_slice(DimensionVec **vecptr, DimensionSlice *slice)
```

Comment:
Can we add an Assert here that the dimension_id is correct?

Reply:
done.